### PR TITLE
[SYCL][CUDA] Add async_group_copy free function implementation

### PR DESCRIPTION
**Pull request by FMarno**
Partial Implement for proposal at #4903.
This adds async_group_copy as a free function, implemented for `group` and `sub_group`.
This also adds optimizations for NVIDIA architectures of sm_80 and over, making it actually async.

Credit to @JackAKirk for most of the implementation.

Tests to follow shortly

### CODE REVIEW

**Comment by JackAKirk** [comment]
Tests are ready to add to the llvm-test-suite here: https://github.com/intel/llvm-test-suite/compare/intel...JackAKirk:async-collective-tests?expand=1

File: sycl/include/sycl/ext/oneapi/group_algorithm.hpp [hunk 1]
```diff
@@ -65,6 +65,125 @@ using EnableIfIsNonNativeOp = cl::sycl::detail::enable_if_t<
...
+} // namespace detail
+
+/// Asynchronously copies a number of elements specified by \p numElements
+/// from the source pointed by \p src to destination pointed by \p dest
+/// with a source stride specified by \p srcStride, and returns a SYCL
+/// device_event which can be used to wait on the completion of the copy.
+/// Permitted types for dataT are all scalar and vector types, except boolean.
+template <typename Group, typename dataT>
+detail::enable_if_t<is_group_v<Group> && !detail::is_bool<dataT>::value,
+                    device_event>
```
Comment (alexbatashev) [inline sycl/include/sycl/ext/oneapi/group_algorithm.hpp:85]: This is C++14 feature, we can use it safely
```suggestion
std::enable_if_t<is_group_v<Group> && !detail::is_bool<dataT>::value,
                    device_event>
```

File: sycl/include/sycl/ext/oneapi/group_algorithm.hpp [hunk 2]
```diff
@@ -65,6 +65,125 @@ using EnableIfIsNonNativeOp = cl::sycl::detail::enable_if_t<
...
+/// with a stride specified by \p Stride, and returns a SYCL device_event
+/// which can be used to wait on the completion of the copy.
+template <typename Group, typename T, access::address_space DestS,
+          access::address_space SrcS>
+detail::enable_if_t<is_group_v<Group> && detail::is_scalar_bool<T>::value,
+                    device_event>
+async_group_copy(Group g, multi_ptr<T, SrcS> Src, multi_ptr<T, DestS> Dest,
+                 size_t NumElements, size_t Stride) {
+  static_assert(sizeof(bool) == sizeof(uint8_t),
+                "Async copy to/from bool memory is not supported.");
```
Comment (alexbatashev) [inline sycl/include/sycl/ext/oneapi/group_algorithm.hpp:128]: I guess the idea here is that `bool` takes 1 byte? Wouldn't it be better then to replace `uint8_t` with `char`?

File: sycl/include/sycl/ext/oneapi/group_algorithm.hpp [hunk 3]
```diff
@@ -65,6 +65,125 @@ using EnableIfIsNonNativeOp = cl::sycl::detail::enable_if_t<
...
+template <typename Group, typename T, access::address_space DestS,
+          access::address_space SrcS>
+detail::enable_if_t<is_group_v<Group> && detail::is_vector_bool<T>::value,
+                    device_event>
+async_group_copy(Group g, multi_ptr<T, SrcS> Src, multi_ptr<T, DestS> Dest,
+                 size_t NumElements, size_t Stride) {
+  static_assert(sizeof(bool) == sizeof(uint8_t),
+                "Async copy to/from bool memory is not supported.");
+  using VecT = detail::change_base_type_t<T, uint8_t>;
+  auto DestP = multi_ptr<VecT, DestS>(reinterpret_cast<VecT *>(Dest.get()));
```
Comment (alexbatashev) [inline sycl/include/sycl/ext/oneapi/group_algorithm.hpp:149]: Same as above. This `reinterpret_cast` is UB. Can we replace `uint8_t` with `char`?

File: sycl/include/sycl/ext/oneapi/group_algorithm.hpp [hunk 4]
```diff
@@ -65,6 +65,125 @@ using EnableIfIsNonNativeOp = cl::sycl::detail::enable_if_t<
...
+}
+
+template <typename Group, typename... eventTN>
+void wait_for(Group g, eventTN... Events) {
+  // Events.wait() calls __spirv_GroupWaitEvents
+  // __spirv_GroupWaitEvents ignores event_list and calls __spirv_ControlBarrier
+  // https://github.com/intel/llvm/blob/sycl/libclc/generic/libspirv/async/wait_group_events.cl
+  // __spirv_ControlBarrier calls __syncthreads or __nvvm_bar_warp_sync
+  // https://github.com/intel/llvm/blob/sycl/libclc/ptx-nvidiacl/libspirv/synchronization/barrier.cl
+  (Events.ext_oneapi_wait(g), ...);
```
Comment (alexbatashev) [inline sycl/include/sycl/ext/oneapi/group_algorithm.hpp:184]: Folding expressions are C++ 17 feature. @romanovvlad should we guard those with macros for extensions?

**Review by alexbatashev — COMMENTED** [review]
just a few nits

File: sycl/include/sycl/ext/oneapi/group_algorithm.hpp (hunk 4 above)
Comment (romanovvlad) [reply sycl/include/sycl/ext/oneapi/group_algorithm.hpp]: It depends, there is a test: sycl/test/basic_tests/stdcpp_compat.cpp. If it passes there is no need to guard.

File: sycl/include/sycl/ext/oneapi/group_algorithm.hpp (hunk 1 above)
Comment (FMarno) [reply sycl/include/sycl/ext/oneapi/group_algorithm.hpp]: done

File: sycl/include/sycl/ext/oneapi/group_algorithm.hpp (hunk 2 above)
Comment (FMarno) [reply sycl/include/sycl/ext/oneapi/group_algorithm.hpp]: Yes, I think that is the idea. I've replaced that now.

File: sycl/include/sycl/ext/oneapi/group_algorithm.hpp (hunk 3 above)
Comment (FMarno) [reply sycl/include/sycl/ext/oneapi/group_algorithm.hpp]: done

File: sycl/include/sycl/ext/oneapi/group_algorithm.hpp (hunk 4 above)
Comment (FMarno) [reply sycl/include/sycl/ext/oneapi/group_algorithm.hpp]: Works fine on my machine

File: libclc/generic/include/clc/async/common.h [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+         __spirv_LocalInvocationId_x();                                        \
+  } else {                                                                     \
+    size = __spirv_SubgroupSize();                                             \
+    id = __spirv_SubgroupLocalInvocationId();                                  \
+  }                                                                            \
+  size_t i;                                                                    \
+                                                                               \
+  for (i = id; i < num_gentypes; i += size) {                                  \
+    dst[i * DST_STRIDE] = src[i * SRC_STRIDE];                                 \
+  }
```
Comment (bader) [inline libclc/generic/include/clc/async/common.h:31]:
```suggestion
                                                                               \
  for (size_t i = id; i < num_gentypes; i += size) {                           \
    dst[i * DST_STRIDE] = src[i * SRC_STRIDE];                                 \
  }
```

File: libclc/ptx-nvidiacl/libspirv/async/async_work_group_strided_copy.cl [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+        id = (__spirv_WorkgroupSize_y() * __spirv_WorkgroupSize_x() *          \
+              __spirv_LocalInvocationId_z()) +                                 \
+             (__spirv_WorkgroupSize_x() * __spirv_LocalInvocationId_y()) +     \
+             __spirv_LocalInvocationId_x();                                    \
+        size = __spirv_WorkgroupSize_x() * __spirv_WorkgroupSize_y() *         \
+               __spirv_WorkgroupSize_z();                                      \
+      } else {                                                                 \
+        size = __spirv_SubgroupSize();                                         \
+        id = __spirv_SubgroupLocalInvocationId();                              \
+      }                                                                        \
```
Comment (bader) [inline libclc/ptx-nvidiacl/libspirv/async/async_work_group_strided_copy.cl:53]: This is exact copy of the code from `libclc/generic/include/clc/async/common.h`. I think it make sense to factor it out into a separate function/macro.

File: libclc/ptx-nvidiacl/libspirv/async/async_work_group_strided_copy.cl [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+      size_t i;                                                                \
+      for (i = id; i < num_gentypes; i += size) {                              \
```
Comment (bader) [inline libclc/ptx-nvidiacl/libspirv/async/async_work_group_strided_copy.cl:55]:
```suggestion
      for (size_t i = id; i < num_gentypes; i += size) {                       \
```

File: libclc/ptx-nvidiacl/libspirv/async/async_work_group_strided_copy.cl [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
```
Comment (bader) [inline libclc/ptx-nvidiacl/libspirv/async/async_work_group_strided_copy.cl:91]:
```suggestion
      for (size_t i = id; i < num_gentypes; i += size) {                       \
```

**Review by bader — APPROVED** [review]

File: libclc/ptx-nvidiacl/libspirv/async/async_work_group_strided_copy.cl (hunk 6 above)
Comment (FMarno) [reply libclc/ptx-nvidiacl/libspirv/async/async_work_group_strided_copy.cl]: sure, updated.

File: libclc/ptx-nvidiacl/libspirv/async/async_work_group_strided_copy.cl (hunk 7 above)
Comment (FMarno) [reply libclc/ptx-nvidiacl/libspirv/async/async_work_group_strided_copy.cl]: done

File: libclc/ptx-nvidiacl/libspirv/async/async_work_group_strided_copy.cl (hunk 8 above)
Comment (FMarno) [reply libclc/ptx-nvidiacl/libspirv/async/async_work_group_strided_copy.cl]: done

File: libclc/generic/include/clc/async/common.h [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+    ID = (__spirv_WorkgroupSize_y() * __spirv_WorkgroupSize_x() *              \
+          __spirv_LocalInvocationId_z()) +                                     \
+         (__spirv_WorkgroupSize_x() * __spirv_LocalInvocationId_y()) +         \
+         __spirv_LocalInvocationId_x();                                        \
+  } else {                                                                     \
+    SIZE = __spirv_SubgroupSize();                                             \
+    ID = __spirv_SubgroupLocalInvocationId();                                  \
+  }
+
+
```
Comment (bader) [inline libclc/generic/include/clc/async/common.h:25]:
```suggestion

```

**Review by bader — DISMISSED** [review]

**Comment by bader** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/552

**Comment by JackAKirk** [comment]
> /verify with [intel/llvm-test-suite#552](https://github.com/intel/llvm-test-suite/pull/552)

I've updated #552 reflecting the latest commits in this PR.

**Review by alexbatashev — DISMISSED** [review]

**Comment by Pennycook** [comment]
I don't think we should merge this before the spec (https://github.com/intel/llvm/pull/4950) is merged, and I don't think we should mark it as a final/implemented/supported extension until we get clarification from Khronos on whether group algorithms act as synchronization points (https://gitlab.khronos.org/sycl/Specification/-/issues/576).

If we want to merge this soon, it may be worth considering putting it in the `experimental::` namespace.  That would make the feature available for testing, and we could take things out of `experimental::` once we've reached consensus on the expected synchronization behavior.  Tagging @gmlueck for comment, since he's also a reviewer on https://github.com/intel/llvm/pull/4950.

**Comment by gmlueck** [comment]
> Moved async_group_copy to experimental

If the implementation is moved to "experimental", then the API spec in #4903 should also be updated to say the extension is experimental.

**Comment by FMarno** [comment]
@gmlueck @Pennycook I've moved the implementation to the experimental namespace and updated the proposal in #4950. We'll update #4903 in the near future.

**Comment by JackAKirk** [comment]
All Opencl types with 4, 8 or 16 byte alignment have been optimized in the cuda backend for sm_80.  The test is up to date here: https://github.com/intel/llvm-test-suite/pull/552

**Review by bader — CHANGES_REQUESTED** [review]
@FMarno, could you resolve merge conflicts, please?

**Comment by JackAKirk** [comment]
> @FMarno, could you resolve merge conflicts, please?

We've marked this as draft for the moment.  The `merge sycl` broke this impl due to some other changes from https://github.com/intel/llvm/commit/8a3bd1c5a9255fc4356f707009c9e5a463aca6ac.  I'll fix it now.

**Pull request closed without merging**